MQ2GMCheck.cpp

```diff
@@ -864,253 +992,261 @@ void GMCheckCmd(PSPAWNINFO pChar, PCHAR szLine)
+			if (szTmp[0] && _FileExists(szTmp))
+				PlayGMSound(szTmp);
+			else
+				PlayGMSound(szEnterSound);
+		}
+
+		if (!bGMQuiet && bGMBeep)
+		{
+			PlaySound(NULL, NULL, SND_NODEFAULT);
+			PlaySound("SystemAsterisk", NULL, SND_ALIAS | SND_ASYNC | SND_NODEFAULT);
+		}
+
+		if (bGMPopup)
+			DisplayOverlayText(szPopup, CONCOLOR_RED, 100, 500, 500, 3000);
 	}
 }
 
-PLUGIN_API VOID OnRemoveSpawn(PSPAWNINFO pSpawn)
+PLUGIN_API VOID OnRemoveSpawn(PlayerClient* pSpawn)
 {
-	char szTmp[MAX_STRING] = { 0 };
-	char szMsg[MAX_STRING], szPopup[MAX_STRING];
-	bool GMFound = false;
-	if (bGMCheck)
+	if (bGMCheck && pSpawn && pSpawn->GM)
 	{
-		if (pSpawn)
+		bool GMFound = false;
+		for (unsigned int x = 0; x < GMNames.size(); x++)
 		{
-			if (pSpawn->GM)
+			std::string& VectorRef = GMNames[x];
+			if (!_stricmp(pSpawn->DisplayedName, VectorRef.c_str()))
 			{
-				for (unsigned int x = 0; x < GMNames.size(); x++)
-				{
-					std::string& VectorRef = GMNames[x];
-					if (!_stricmp(pSpawn->DisplayedName, VectorRef.c_str()))
-					{
-						GMNames.erase(GMNames.begin() + x);
-						GMFound = true;
-					}
-				}
-				if (!GMFound)
-					return;
-				sprintf_s(szMsg, "\agGM %s \ayhas left the zone at \ag%s", pSpawn->DisplayedName, DisplayTime());
-				sprintf_s(szPopup, "GM %s has left the zone at %s", pSpawn->DisplayedName, DisplayTime());
-				if (bGMChatAlert)
-					WriteChatf(szMsg);
-				if (!(GMNames.size()))
-				{
-					if (bGMCmdActive)
-					{
-						strcpy_s(szTmp, szGMEnterCmdIf);
-						if (MCEval(szTmp))
-						{
-							if (szGMLeaveCmd[0])
-							{
-								if (szGMLeaveCmd[0] == '/')
-								{
-									DoCommand((GetCharInfo() && GetCharInfo()->pSpawn) ? GetCharInfo()->pSpawn : NULL, szGMLeaveCmd);
-									bGMCmdActive = false;
-								}
-							}
-						}
-					}
-				}
-				if (!bGMQuiet)
-				{
-					if (bGMSound)
-					{
-						GetPrivateProfileString(pSpawn->DisplayedName, "LeaveSound", "", szTmp, MAX_STRING, INIFileName);
-						if (szTmp[0] && _FileExists(szTmp))
-							PlayGMSound(szTmp);
-						else
-							PlayGMSound(szLeaveSound);
-					}
-					if (bGMBeep)
-					{
-						PlaySound(NULL, NULL, SND_NODEFAULT);
-						PlaySound("SystemDefault", NULL, SND_ALIAS | SND_ASYNC | SND_NODEFAULT);
-					}
-				}
-				if (bGMPopup)
-					DisplayOverlayText(szPopup, CONCOLOR_GREEN, 100, 500, 500, 3000);
+				GMNames.erase(GMNames.begin() + x);
+				GMFound = true;
+			}
+		}
+
+		if (!GMFound)
+			return;
+
+		if (bGMChatAlert)
+		{
+			char szMsg[MAX_STRING] = { 0 };
+			sprintf_s(szMsg, "\agGM %s \ayhas left the zone at \ag%s", pSpawn->DisplayedName, DisplayTime());
+			WriteChatf("%s%s", PluginMsg, szMsg);
+		}
+
+		if (GMNames.empty() && bGMCmdActive)
+		{
+			char szTmp[MAX_STRING] = { 0 };
+			strcpy_s(szTmp, szGMLeaveCmdIf);
+			if (MCEval(szTmp) && szGMLeaveCmd[0] && szGMLeaveCmd[0] == '/')
+			{
+				DoCommand((GetCharInfo() && GetCharInfo()->pSpawn) ? GetCharInfo()->pSpawn : NULL, szGMLeaveCmd);
```

Comment:
what is even happening here lol.

`EzCommand(szGMLeaveCmd)`

Reply:
From the looks of it, the author of this line of code intended to use the DoCommand and pass the spawn but only if the spawn was valid, otherwise they would send it NULL. Which iirc used to work to use NULL but doesn't anymore. Can't be sure.

Reply:
the suggested change is equivalent. GetCharInfo()->pSpawn is equivalent to pLocalPlayer